### PR TITLE
Memory/reduce allocations in peer manager

### PR DESCRIPTION
**Pull request by LukaszRozmej**
## Changes:
Reduce allocations in peer manager
More context: https://www.notion.so/nethermind/Memory-investigations-2aa0fa86988c4bf88849a66b7328f203#60e7fde6cfee4b5c9daa3e4b8507196f

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Other (please describe): memory fragmentation and performance

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by kkokosa — COMMENTED** [review]
Looks good to me. Reduces time 3x and allocations 4x:

```
|                  Method |       Mean |    Error |   StdDev |  Gen 0 | Allocated |
|------------------------ |-----------:|---------:|---------:|-------:|----------:|
| CleanupCandidatePeers_1 | 1,126.9 ns | 22.51 ns | 24.08 ns | 0.0992 |   1,664 B |
| CleanupCandidatePeers_2 |   406.7 ns |  6.85 ns |  6.41 ns | 0.0262 |     440 B
```

Allocations are from various places:

image

**Review by tanishqjasoria — APPROVED** [review]
Client instances seems to be running fine.

File: src/Nethermind/Nethermind.Network/PeerManager.cs [hunk 1]
```diff
@@ -78,12 +81,13 @@ public class PeerManager : IPeerManager
             _stats = stats ?? throw new ArgumentNullException(nameof(stats));
             _networkConfig = networkConfig ?? throw new ArgumentNullException(nameof(networkConfig));
             _peerPool = peerPool;
+            _candidates = new List<PeerStats>(networkConfig.MaxActivePeers * 2);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/PeerManager.cs:84]: just pre-allocate  space

File: src/Nethermind/Nethermind.Network/PeerManager.cs [hunk 2]
```diff
@@ -459,11 +463,10 @@ private void SelectAndRankCandidates()
             }
 
             bool hasOnlyStaticNodes = false;
-            List<Peer> staticPeers = _peerPool.StaticPeers;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/PeerManager.cs:462]: avoid allocating list

File: src/Nethermind/Nethermind.Network/PeerManager.cs [hunk 3]
```diff
@@ -843,39 +846,51 @@ private void StopTimers()
...
-            int countToRemove = candidates.Count - _networkConfig.MaxCandidatePeerCount;
-            Peer[] failedValidationCandidates = candidates.Where(x => _stats.HasFailedValidation(x.Node))
-                .OrderBy(x => _stats.GetCurrentReputation(x.Node)).ToArray();
-            Peer[] otherCandidates = candidates.Except(failedValidationCandidates).Except(_peerPool.ActivePeers.Values).OrderBy(x => _stats.GetCurrentReputation(x.Node)).ToArray();
-            Peer[] nodesToRemove = failedValidationCandidates.Length <= countToRemove
-                ? failedValidationCandidates
-                : failedValidationCandidates.Take(countToRemove).ToArray();
-            int failedValidationRemovedCount = nodesToRemove.Length;
-            int remainingCount = countToRemove - failedValidationRemovedCount;
-            if (remainingCount > 0)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/PeerManager.cs:862]: Its equivalent to just sorting by: HasFailedValidation desc first and GetCurrentReputation asc second!

File: src/Nethermind/Nethermind.Network/PeerPool.cs [hunk 4]
```diff
@@ -35,24 +35,24 @@ public class PeerPool : IPeerPool
...
         public ConcurrentDictionary<PublicKey, Peer> Peers { get; } = new();
         private readonly ConcurrentDictionary<PublicKey, Peer> _staticPeers = new();
-        
-        public List<Peer> NonStaticPeers => Peers.Values.Where(p => !p.Node.IsStatic).ToList();
-        public List<Peer> StaticPeers => _staticPeers.Values.ToList();
-        
+
+        public IEnumerable<Peer> NonStaticPeers => Peers.Values.Where(p => !p.Node.IsStatic);
+        public IEnumerable<Peer> StaticPeers => _staticPeers.Values;
+
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/PeerPool.cs:51]: avoid allocations, ConcurrentDictionary enumerator is thread-safe

**Review by LukaszRozmej — COMMENTED** [review]
Adding comments that help with review

**Review by yaziciahmet — APPROVED** [review]

File: src/Nethermind/Nethermind.Network/PeerManager.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/PeerManager.cs:84]: _candidates are later used only by one thread so no thread-safety measures required

File: src/Nethermind/Nethermind.Network/PeerManager.cs [hunk 5]
```diff
@@ -843,39 +846,51 @@ private void StopTimers()
...
-                return;
-            }
+            int peerCount = _peerPool.PeerCount;
 
-            // may further optimize allocations here
-            List<Peer> candidates = _peerPool.NonStaticPeers;
-            int countToRemove = candidates.Count - _networkConfig.MaxCandidatePeerCount;
-            Peer[] failedValidationCandidates = candidates.Where(x => _stats.HasFailedValidation(x.Node))
-                .OrderBy(x => _stats.GetCurrentReputation(x.Node)).ToArray();
-            Peer[] otherCandidates = candidates.Except(failedValidationCandidates).Except(_peerPool.ActivePeers.Values).OrderBy(x => _stats.GetCurrentReputation(x.Node)).ToArray();
```
Comment (dceleda) [inline src/Nethermind/Nethermind.Network/PeerManager.cs:856]: _Except(_peerPool.ActivePeers.Values)_ condition has been removed. Are we ok with that?

File: src/Nethermind/Nethermind.Network/PeerManager.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/PeerManager.cs:462]: I use it directly later, as an IEnumerable. I avoid allocating a List.

File: src/Nethermind/Nethermind.Network/PeerManager.cs (hunk 5 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/PeerManager.cs:856]: Good catch. I think I missed it! Those need to be sorted as last items and later removed from the collection!

This is very important for this PR!

File: src/Nethermind/Nethermind.Network/PeerManager.cs [hunk 6]
```diff
@@ -843,39 +846,51 @@ private void StopTimers()
...
+                return;
             }
 
-            if (nodesToRemove.Length > 0)
+            try
             {
-                _logger.Info($"Removing {nodesToRemove.Length} out of {candidates.Count} peer candidates (candidates cleanup).");
-                foreach (Peer peer in nodesToRemove)
+                int failedValidationCandidatesCount = 0;
+                foreach ((_, Peer peer) in _peerPool.Peers)
```
[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/PeerManager.cs:859]: Yes its currently not used I left it for completness (we have StaticPeers)

File: src/Nethermind/Nethermind.Network/PeerManager.cs [hunk 7]
```diff
@@ -886,5 +901,32 @@ private enum ActivePeerSelectionCounter
...
+
+            public PeerStats(Peer peer, bool failedValidation, long currentReputation)
+            {
+                Peer = peer;
+                FailedValidation = failedValidation;
+                CurrentReputation = currentReputation;
+            }
+        }
+
+        private class PeerStatsComparer : IComparer<PeerStats>
```
[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/PeerManager.cs:919]: Both are non-allocating I prefer to have it under IComparer

File: src/Nethermind/Nethermind.Network/PeerManager.cs (hunk 5 above)
Comment (dceleda) [reply src/Nethermind/Nethermind.Network/PeerManager.cs:856]: Could we add a unit test for that to avoid breaking the condition again in the future?

**Pull request merged**